hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,516 @@
+	}
+
+	t := v.Type()
+	controllerName := fmt.Sprintf("%sController", t.Name())
+
+	// Use the provided GVK to construct a new runtime object of the desired concrete type.
+	gvk, err := apiutil.GVKForObject(obj, mgr.GetScheme())
+	if err != nil {
+		return errors.Wrapf(err, "couldn't create GVK for obj %T", obj)
+	}
+	log.V(4).Info("Registering", "GVK", gvk)
+
+	// TODO: Do we need to add any index fields here? DavidJ's controller index's status.id - see its usage
+	// TODO: of IndexField
+
+	kubeClient := kubeclient.NewClient(mgr.GetClient(), mgr.GetScheme())
+
+	reconciler := &GenericReconciler{
+		ARMClient:        applier,
+		KubeClient:       kubeClient,
+		ResourceResolver: armresourceresolver.NewResolver(kubeClient),
+		Name:             t.Name(),
+		Log:              log.WithName(controllerName),
+		Recorder:         mgr.GetEventRecorderFor(controllerName),
+		GVK:              gvk,
+	}
+
+	ctrlBuilder := ctrl.NewControllerManagedBy(mgr).
+		For(obj).
+		WithOptions(options)
+
+	c, err := ctrlBuilder.Build(reconciler)
+	if err != nil {
+		return errors.Wrap(err, "unable to build controllers / reconciler")
+	}
+
+	reconciler.Controller = c
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(obj).
+		Complete()
+}
+
+// Reconcile will take state in K8s and apply it to Azure
+func (gr *GenericReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := gr.Log.WithValues("name", req.Name, "namespace", req.Namespace)
+
+	obj, err := gr.KubeClient.GetObject(ctx, req.NamespacedName, gr.GVK)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if obj == nil {
+		// This means that the resource doesn't exist
+		return ctrl.Result{}, nil
+	}
+
+	// Always operate on a copy rather than the object from the client, as per
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md, which says:
+	// Never mutate original objects! Caches are shared across controllers, this means that if you mutate your "copy"
+	// (actually a reference or shallow copy) of an object, you'll mess up other controllers (not just your own).
+	obj = obj.DeepCopyObject()
+
+	// The Go type for the Kubernetes object must understand how to
+	// convert itself to/from the corresponding Azure types.
+	metaObj, ok := obj.(genruntime.MetaObject)
+	if !ok {
+		return ctrl.Result{}, errors.Errorf("object is not a genruntime.MetaObject: %+v - type: %T", obj, obj)
+	}
+
+	objWrapper := NewReconcileMetadata(metaObj, log)
+	action, actionFunc, err := gr.DetermineReconcileAction(objWrapper)
+
+	if err != nil {
+		log.Error(err, "error determining reconcile action")
+		gr.Recorder.Event(metaObj, v1.EventTypeWarning, "DetermineReconcileActionError", err.Error())
+	}
+
+	result, err := actionFunc(ctx, action, objWrapper)
+	if err != nil {
+		log.Error(err, "Error during reconcile", "action", action)
+		gr.Recorder.Event(metaObj, v1.EventTypeWarning, "ReconcileActionError", err.Error())
+	}
```

Comment:
Added this for visual similarity with the err check above.

BTW I think that we should consider find an interface we can collapse log + recorder into so that they're one thing and a bit cleaner... but that's for another time.